File: src/types/workflow.ts
/**
 * Workflow schema type definitions
 *
 * WorkflowSchema defines how external columns map to internal statuses,
 * enabling custom kanban workflows while preserving state machine behavior.
 *
 * Note: Backend outputs snake_case (Rust default). Response schemas validate
 * the raw data, then transform functions convert to camelCase for frontend.
 */

import { z } from "zod";
import { InternalStatusSchema, type InternalStatus } from "./status";

// ============================================
// External Sync Configuration Types (future)
// ============================================

/**
 * Supported external sync providers
 */
export const SyncProviderSchema = z.enum(["jira", "github", "linear", "notion"]);
export type SyncProvider = z.infer<typeof SyncProviderSchema>;

/**
 * All sync provider values as a readonly array
 */
export const SYNC_PROVIDER_VALUES = SyncProviderSchema.options;

/**
 * Sync direction options
 */
export const SyncDirectionSchema = z.enum(["pull", "push", "bidirectional"]);
export type SyncDirection = z.infer<typeof SyncDirectionSchema>;

/**
 * All sync direction values as a readonly array
 */
export const SYNC_DIRECTION_VALUES = SyncDirectionSchema.options;

/**
 * Conflict resolution strategy
 */
export const ConflictResolutionSchema = z.enum([
  "external_wins",
  "internal_wins",
  "manual",
]);
export type ConflictResolution = z.infer<typeof ConflictResolutionSchema>;

/**
 * All conflict resolution values as a readonly array
 */
export const CONFLICT_RESOLUTION_VALUES = ConflictResolutionSchema.options;

/**
 * Mapping from an external status to internal status
 */
export const ExternalStatusMappingSchema = z.object({
  /** The external status name from the provider */
  externalStatus: z.string(),
  /** The internal status to map to */
  internalStatus: InternalStatusSchema,
  /** The workflow column to display in */
  columnId: z.string(),
});
export type ExternalStatusMapping = z.infer<typeof ExternalStatusMappingSchema>;

/**
 * Sync direction settings
 */
export const SyncSettingsSchema = z.object({
  /** Sync direction */
  direction: SyncDirectionSchema,
  /** Enable webhook for real-time sync */
  webhook: z.boolean().optional(),
});
export type SyncSettings = z.infer<typeof SyncSettingsSchema>;

/**
 * External sync configuration (placeholder for future implementation)
 */
export const ExternalSyncConfigSchema = z.object({
  /** The external provider */
  provider: SyncProviderSchema,
  /** Status mapping from external to internal (keyed by external status) */
  mapping: z.record(z.string(), ExternalStatusMappingSchema).default({}),
  /** Sync settings */
  sync: SyncSettingsSchema,
  /** How to resolve conflicts */
  conflictResolution: ConflictResolutionSchema,
});
export type ExternalSyncConfig = z.infer<typeof ExternalSyncConfigSchema>;

// ============================================
// State Grouping Types (Multi-State Columns)
// ============================================

/**
 * State group within a column
 * Allows multiple internal statuses to be grouped and displayed within a single column
 */
export const StateGroupSchema = z.object({
  /** Unique group identifier within the column */
  id: z.string(),
  /** Display label for the group header (e.g., "Fresh Tasks", "Needs Revision") */
  label: z.string(),
  /** Internal statuses that belong to this group */
  statuses: z.array(InternalStatusSchema),
  /** Optional Lucide icon name for the group */
  icon: z.string().optional(),
  /** Optional accent color for the group (CSS color value) */
  accentColor: z.string().optional(),
  /** Whether tasks can be dragged FROM this group (default: true) */
  canDragFrom: z.boolean().optional(),
  /** Whether tasks can be dropped TO this group (default: true) */
  canDropTo: z.boolean().optional(),
});

export type StateGroup = z.infer<typeof StateGroupSchema>;

// ============================================
// Column Behavior and Workflow Types
// ============================================

/**
 * Column behavior configuration
 * Controls how tasks behave when moved to this column
 */
const ColumnBehaviorSchema = z.object({
  /** Skip AI review for tasks in this column */
  skipReview: z.boolean().optional(),
  /** Automatically advance to next column when execution completes */
  autoAdvance: z.boolean().optional(),
  /** Agent profile to use for tasks in this column */
  agentProfile: z.string().optional(),
});

/**
 * Schema for a workflow column
 * Maps an external column to an internal status
 */
export const WorkflowColumnSchema = z.object({
  /** Unique column identifier */
  id: z.string(),
  /** Display name for the column */
  name: z.string(),
  /** Optional color for the column header */
  color: z.string().optional(),
  /** Optional icon name for the column */
  icon: z.string().optional(),
  /** Internal status this column maps to (primary status for single-state columns) */
  mapsTo: InternalStatusSchema,
  /** Optional behavior configuration */
  behavior: ColumnBehaviorSchema.optional(),
  /** Optional state groups for multi-state columns */
  groups: z.array(StateGroupSchema).optional(),
});

export type WorkflowColumn = z.infer<typeof WorkflowColumnSchema>;

/**
 * Default agent profiles for the workflow
 */
const WorkflowDefaultsSchema = z.object({
  /** Default worker agent profile */
  workerProfile: z.string().optional(),
  /** Default reviewer agent profile */
  reviewerProfile: z.string().optional(),
});

/**
 * Schema for a complete workflow definition
 * Note: Named WorkflowSchemaZ to avoid collision with the TypeScript type
 */
export const WorkflowSchemaZ = z.object({
  /** Unique workflow identifier */
  id: z.string(),
  /** Display name for the workflow */
  name: z.string(),
  /** Optional description */
  description: z.string().optional(),
  /** Ordered list of columns in the workflow */
  columns: z.array(WorkflowColumnSchema),
  /** Optional external sync configuration (future implementation) */
  externalSync: ExternalSyncConfigSchema.optional(),
  /** Optional default agent profiles */
  defaults: WorkflowDefaultsSchema.optional(),
  /** Whether this is the default workflow */
  isDefault: z.boolean().default(false),
});

export type WorkflowSchema = z.infer<typeof WorkflowSchemaZ>;

// ============================================
// Response Schemas (snake_case from Rust backend)
// ============================================

/**
 * State group response schema (snake_case from Rust)
 */
export const StateGroupResponseSchema = z.object({
  id: z.string(),
  label: z.string(),
  statuses: z.array(z.string()),
  icon: z.string().optional(),
  accent_color: z.string().optional(),
  can_drag_from: z.boolean().optional(),
  can_drop_to: z.boolean().optional(),
});

/**
 * Workflow column response schema (snake_case from Rust)
 */
export const WorkflowColumnResponseSchema = z.object({
  id: z.string(),
  name: z.string(),
  maps_to: z.string(),
  color: z.string().optional(),
  icon: z.string().optional(),
  skip_review: z.boolean().optional(),
  auto_advance: z.boolean().optional(),
  agent_profile: z.string().optional(),
  groups: z.array(StateGroupResponseSchema).optional(),
});

/**
 * Workflow response schema (snake_case from Rust)
 */
export const WorkflowResponseSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().optional(),
  columns: z.array(WorkflowColumnResponseSchema),
  is_default: z.boolean(),
  worker_profile: z.string().optional(),
  reviewer_profile: z.string().optional(),
});

// ============================================
// Transform Functions
// ============================================

/**
 * Transform state group from snake_case response to camelCase frontend type
 */
export function transformStateGroup(
  raw: z.infer<typeof StateGroupResponseSchema>
): StateGroup {
  return {
    id: raw.id,
    label: raw.label,
    statuses: raw.statuses as InternalStatus[],
    icon: raw.icon,
    accentColor: raw.accent_color,
    canDragFrom: raw.can_drag_from,
    canDropTo: raw.can_drop_to,
  };
}

/**
 * Transform workflow column from snake_case response to camelCase frontend type
 */
export function transformWorkflowColumn(
  raw: z.infer<typeof WorkflowColumnResponseSchema>
): WorkflowColumn {
  return {
    id: raw.id,
    name: raw.name,
    mapsTo: raw.maps_to as InternalStatus,
    color: raw.color,
    icon: raw.icon,
    behavior: (raw.skip_review !== undefined || raw.auto_advance !== undefined || raw.agent_profile !== undefined)
      ? {
          skipReview: raw.skip_review,
          autoAdvance: raw.auto_advance,
          agentProfile: raw.agent_profile,
        }
      : undefined,
    groups: raw.groups?.map(transformStateGroup),
  };
}

/**
 * Transform workflow from snake_case response to camelCase frontend type
 */
export function transformWorkflow(
  raw: z.infer<typeof WorkflowResponseSchema>
): WorkflowSchema {
  return {
    id: raw.id,
    name: raw.name,
    description: raw.description,
    columns: raw.columns.map(transformWorkflowColumn),
    isDefault: raw.is_default,
    defaults: (raw.worker_profile !== undefined || raw.reviewer_profile !== undefined)
      ? {
          workerProfile: raw.worker_profile,
          reviewerProfile: raw.reviewer_profile,
        }
      : undefined,
  };
}

/**
 * Default RalphX workflow with 5 columns
 * Maps to the standard kanban board structure
 * Multi-state columns use groups to provide visibility into task state
 */
export const defaultWorkflow: WorkflowSchema = {
  id: "ralphx-default",
  name: "RalphX Default",
  description: "Standard kanban workflow for AI-driven development",
  columns: [
    { id: "draft", name: "Draft", mapsTo: "backlog" },
    {
      id: "ready",
      name: "Ready",
      mapsTo: "ready",
      groups: [
        {
          id: "fresh",
          label: "Fresh Tasks",
          statuses: ["ready"],
          canDragFrom: true,
          canDropTo: true,
        },
        {
          id: "needs_revision",
          label: "Needs Revision",
          statuses: ["revision_needed"],
          icon: "RotateCcw",
          accentColor: "hsl(var(--warning))",
          canDragFrom: true,
          canDropTo: false, // Only review process can add here
        },
      ],
    },
    {
      id: "in_progress",
      name: "In Progress",
      mapsTo: "executing",
      groups: [
        {
          id: "first_attempt",
          label: "First Attempt",
          statuses: ["executing"],
          canDragFrom: false, // System-managed (agent working)
          canDropTo: false,
        },
        {
          id: "revising",
          label: "Revising",
          statuses: ["re_executing"],
          icon: "RefreshCw",
          accentColor: "hsl(var(--warning))",
          canDragFrom: false, // System-managed (agent revising)
          canDropTo: false,
        },
      ],
    },
    {
      id: "in_review",
      name: "In Review",
      mapsTo: "pending_review",
      groups: [
        {
          id: "waiting_ai",
          label: "Waiting for AI",
          statuses: ["pending_review"],
          icon: "Clock",
          canDragFrom: false, // System-managed
          canDropTo: false,
        },
        {
          id: "ai_reviewing",
          label: "AI Reviewing",
          statuses: ["reviewing"],
          icon: "Bot",
          accentColor: "hsl(var(--primary))",
          canDragFrom: false, // System-managed (AI working)
          canDropTo: false,
        },
        {
          id: "ready_approval",
          label: "Ready for Approval",
          statuses: ["review_passed"],
          icon: "CheckCircle",
          accentColor: "hsl(var(--success))",
          canDragFrom: false, // User interacts via Approve/Revise buttons
          canDropTo: false,
        },
        {
          id: "escalated",
          label: "Escalated",
          statuses: ["escalated"],
          icon: "AlertTriangle",
          accentColor: "hsl(var(--warning))",
          canDragFrom: false, // User interacts via Approve/Request Changes buttons
          canDropTo: false,
        },
      ],
    },
    { id: "done", name: "Done", mapsTo: "approved" },
  ],
  isDefault: true,
};

/**
 * Jira-compatible workflow with 5 columns
 * Matches familiar Jira board structure with external sync config
 */
export const jiraCompatibleWorkflow: WorkflowSchema = {
  id: "jira-compat",
  name: "Jira Compatible",
  description: "Jira-style workflow with familiar columns",
  columns: [
    { id: "backlog", name: "Backlog", mapsTo: "backlog" },
    { id: "selected", name: "Selected for Dev", mapsTo: "ready" },
    { id: "in_progress", name: "In Progress", mapsTo: "executing" },
    { id: "in_qa", name: "In QA", mapsTo: "pending_review" },
    { id: "done", name: "Done", mapsTo: "approved" },
  ],
  externalSync: {
    provider: "jira",
    mapping: {},
    sync: {
      direction: "bidirectional",
      webhook: true,
    },
    conflictResolution: "external_wins",
  },
  isDefault: false,
};

/**
 * Built-in workflows array for easy iteration
 */
export const BUILTIN_WORKFLOWS: readonly WorkflowSchema[] = [
  defaultWorkflow,
  jiraCompatibleWorkflow,
] as const;

/**
 * Get a built-in workflow by ID
 */
export function getBuiltinWorkflow(id: string): WorkflowSchema | undefined {
  return BUILTIN_WORKFLOWS.find((w) => w.id === id);
}

File: src-tauri/crates/ralphx-domain/src/entities/workflow.rs
// Workflow entities for custom workflow schemas
// These enable custom Kanban columns that map to internal statuses

use serde::{Deserialize, Serialize};
use std::fmt;
use std::str::FromStr;

use super::status::InternalStatus;

// ============================================
// State Grouping Types (Multi-State Columns)
// ============================================

/// State group within a column
/// Allows multiple internal statuses to be grouped and displayed within a single column
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct StateGroup {
    /// Unique group identifier within the column
    pub id: String,
    /// Display label for the group header (e.g., "Fresh Tasks", "Needs Revision")
    pub label: String,
    /// Internal statuses that belong to this group
    pub statuses: Vec<InternalStatus>,
    /// Optional Lucide icon name for the group
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon: Option<String>,
    /// Optional accent color for the group (CSS color value)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accent_color: Option<String>,
    /// Whether tasks can be dragged FROM this group (default: true)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub can_drag_from: Option<bool>,
    /// Whether tasks can be dropped TO this group (default: true)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub can_drop_to: Option<bool>,
}

impl StateGroup {
    /// Creates a new state group with the given id, label, and statuses
    pub fn new(
        id: impl Into<String>,
        label: impl Into<String>,
        statuses: Vec<InternalStatus>,
    ) -> Self {
        Self {
            id: id.into(),
            label: label.into(),
            statuses,
            icon: None,
            accent_color: None,
            can_drag_from: None,
            can_drop_to: None,
        }
    }

    /// Sets the icon
    pub fn with_icon(mut self, icon: impl Into<String>) -> Self {
        self.icon = Some(icon.into());
        self
    }

    /// Sets the accent color
    pub fn with_accent_color(mut self, color: impl Into<String>) -> Self {
        self.accent_color = Some(color.into());
        self
    }

    /// Sets whether tasks can be dragged from this group
    pub fn with_can_drag_from(mut self, can_drag: bool) -> Self {
        self.can_drag_from = Some(can_drag);
        self
    }

    /// Sets whether tasks can be dropped to this group
    pub fn with_can_drop_to(mut self, can_drop: bool) -> Self {
        self.can_drop_to = Some(can_drop);
        self
    }

    /// Convenience method to lock the group (no drag, no drop)
    /// Use for system-managed states where users should not manually move tasks
    pub fn locked(self) -> Self {
        self.with_can_drag_from(false).with_can_drop_to(false)
    }
}

/// A unique identifier for a Workflow
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct WorkflowId(pub String);

impl WorkflowId {
    /// Creates a new WorkflowId with a random UUID v4
    pub fn new() -> Self {
        Self(uuid::Uuid::new_v4().to_string())
    }

    /// Creates a WorkflowId from an existing string
    pub fn from_string(s: impl Into<String>) -> Self {
        Self(s.into())
    }

    /// Returns the inner string value
    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl Default for WorkflowId {
    fn default() -> Self {
        Self::new()
    }
}

impl fmt::Display for WorkflowId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// A workflow schema defines the Kanban board layout with custom columns
/// that map to internal statuses for consistent side effects
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WorkflowSchema {
    /// Unique identifier
    pub id: WorkflowId,
    /// Display name for the workflow
    pub name: String,
    /// Optional description
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Ordered list of columns in the Kanban board
    pub columns: Vec<WorkflowColumn>,
    /// External sync configuration (future implementation)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub external_sync: Option<ExternalSyncConfig>,
    /// Default agent profile overrides
    #[serde(default)]
    pub defaults: WorkflowDefaults,
    /// Whether this is the default workflow
    #[serde(default)]
    pub is_default: bool,
}

impl WorkflowSchema {
    /// Creates a new workflow schema with the given name and columns
    pub fn new(name: impl Into<String>, columns: Vec<WorkflowColumn>) -> Self {
        Self {
            id: WorkflowId::new(),
            name: name.into(),
            description: None,
            columns,
            external_sync: None,
            defaults: WorkflowDefaults::default(),
            is_default: false,
        }
    }

    /// Sets the description
    pub fn with_description(mut self, description: impl Into<String>) -> Self {
        self.description = Some(description.into());
        self
    }

    /// Sets this as the default workflow
    pub fn as_default(mut self) -> Self {
        self.is_default = true;
        self
    }

    /// Returns the default RalphX workflow
    /// Multi-state columns use groups to provide visibility into task state
    pub fn default_ralphx() -> Self {
        Self {
            id: WorkflowId::from_string("ralphx-default"),
            name: "RalphX Default".to_string(),
            description: Some("Default RalphX workflow".to_string()),
            columns: vec![
                WorkflowColumn::new("draft", "Draft", InternalStatus::Backlog),
                WorkflowColumn::new("ready", "Ready", InternalStatus::Ready).with_groups(vec![
                    StateGroup::new("fresh", "Fresh Tasks", vec![InternalStatus::Ready])
                        .with_can_drag_from(true)
                        .with_can_drop_to(true),
                    StateGroup::new(
                        "needs_revision",
                        "Needs Revision",
                        vec![InternalStatus::RevisionNeeded],
                    )
                    .with_icon("RotateCcw")
                    .with_accent_color("hsl(var(--warning))")
                    .with_can_drag_from(true)
                    .with_can_drop_to(false), // Only review process can add here
                    StateGroup::new("blocked", "Blocked", vec![InternalStatus::Blocked])
                        .with_icon("Ban")
                        .with_accent_color("hsl(var(--warning))")
                        .with_can_drag_from(true)
                        .with_can_drop_to(true),
                    StateGroup::new("paused", "Paused", vec![InternalStatus::Paused])
                        .with_icon("Pause")
                        .with_accent_color("hsl(var(--warning))")
                        .locked(),
                ]),
                WorkflowColumn::new("in_progress", "In Progress", InternalStatus::Executing)
                    .with_groups(vec![
                        StateGroup::new(
                            "first_attempt",
                            "First Attempt",
                            vec![InternalStatus::Executing],
                        )
                        .locked(), // System-managed (agent working)
                        StateGroup::new("revising", "Revising", vec![InternalStatus::ReExecuting])
                            .with_icon("RefreshCw")
                            .with_accent_color("hsl(var(--warning))")
                            .locked(), // System-managed (agent revising)
                    ]),
                WorkflowColumn::new("in_review", "In Review", InternalStatus::PendingReview)
                    .with_groups(vec![
                        StateGroup::new(
                            "waiting_ai",
                            "Waiting for AI",
                            vec![InternalStatus::PendingReview],
                        )
                        .with_icon("Clock")
                        .locked(), // System-managed
                        StateGroup::new(
                            "ai_reviewing",
                            "AI Reviewing",
                            vec![InternalStatus::Reviewing],
                        )
                        .with_icon("Bot")
                        .with_accent_color("hsl(var(--primary))")
                        .locked(), // System-managed (AI working)
                        StateGroup::new(
                            "ready_approval",
                            "Ready for Approval",
                            vec![InternalStatus::ReviewPassed],
                        )
                        .with_icon("CheckCircle")
                        .with_accent_color("hsl(var(--success))")
                        .locked(), // User interacts via Approve/Revise buttons
                        StateGroup::new("escalated", "Escalated", vec![InternalStatus::Escalated])
                            .with_icon("AlertTriangle")
                            .with_accent_color("hsl(var(--warning))")
                            .locked(),
                    ]),
                WorkflowColumn::new("done", "Done", InternalStatus::Approved).with_groups(vec![
                    StateGroup::new(
                        "merging",
                        "Merging",
                        vec![
                            InternalStatus::PendingMerge,
                            InternalStatus::Merging,
                            InternalStatus::WaitingOnPr,
                        ],
                    )
                    .with_icon("GitMerge")
                    .locked(),
                    StateGroup::new(
                        "needs_attention",
                        "Escalated",
                        vec![InternalStatus::MergeIncomplete, InternalStatus::MergeConflict],
                    )
                    .with_icon("AlertTriangle")
                    .with_accent_color("hsl(var(--warning))")
                    .locked(),
                    StateGroup::new(
                        "completed",
                        "Completed",
                        vec![InternalStatus::Merged, InternalStatus::Approved],
                    )
                    .with_icon("CheckCircle")
                    .with_accent_color("hsl(var(--success))")
                    .locked(),
                    StateGroup::new("cancelled", "Cancelled", vec![InternalStatus::Cancelled])
                        .with_icon("XCircle")
                        .locked(),
                    StateGroup::new("failed", "Failed", vec![InternalStatus::Failed])
                        .with_icon("XOctagon")
                        .with_accent_color("hsl(var(--destructive))")
                        .locked(),
                    StateGroup::new("stopped", "Stopped", vec![InternalStatus::Stopped])
                        .with_icon("StopCircle")
                        .with_accent_color("hsl(var(--destructive))")
                        .locked(),
                ]),
            ],
            external_sync: None,
            defaults: WorkflowDefaults::default(),
            is_default: true,
        }
    }

    /// Returns the Jira-compatible workflow
    pub fn jira_compatible() -> Self {
        Self {
            id: WorkflowId::from_string("jira-compat"),
            name: "Jira Compatible".to_string(),
            description: Some("Jira-style workflow with familiar columns".to_string()),
            columns: vec![
                WorkflowColumn::new("backlog", "Backlog", InternalStatus::Backlog),
                WorkflowColumn::new("selected", "Selected for Dev", InternalStatus::Ready),
                WorkflowColumn::new("in_progress", "In Progress", InternalStatus::Executing),
                WorkflowColumn::new("in_qa", "In QA", InternalStatus::PendingReview),
                WorkflowColumn::new("done", "Done", InternalStatus::Approved),
            ],
            external_sync: Some(ExternalSyncConfig {
                provider: SyncProvider::Jira,
                mapping: std::collections::HashMap::new(),
                sync: SyncSettings {
                    direction: SyncDirection::Bidirectional,
                    webhook: Some(true),
                },
                conflict_resolution: ConflictResolution::ExternalWins,
            }),
            defaults: WorkflowDefaults::default(),
            is_default: false,
        }
    }
}

/// A column in a workflow's Kanban board
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WorkflowColumn {
    /// Unique identifier within the workflow
    pub id: String,
    /// Display name for the column
    pub name: String,
    /// Optional color (hex format, e.g., "#ff6b35")
    #[serde(skip_serializing_if = "Option::is_none")]
    pub color: Option<String>,
    /// Optional icon name
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon: Option<String>,
    /// The internal status this column maps to for side effects (primary status)
    pub maps_to: InternalStatus,
    /// Optional behavior overrides for this column
    #[serde(skip_serializing_if = "Option::is_none")]
    pub behavior: Option<ColumnBehavior>,
    /// Optional state groups for multi-state columns
    #[serde(skip_serializing_if = "Option::is_none")]
    pub groups: Option<Vec<StateGroup>>,
}

impl WorkflowColumn {
    /// Creates a new workflow column
    pub fn new(id: impl Into<String>, name: impl Into<String>, maps_to: InternalStatus) -> Self {
        Self {
            id: id.into(),
            name: name.into(),
            color: None,
            icon: None,
            maps_to,
            behavior: None,
            groups: None,
        }
    }

    /// Sets the column color
    pub fn with_color(mut self, color: impl Into<String>) -> Self {
        self.color = Some(color.into());
        self
    }

    /// Sets the column icon
    pub fn with_icon(mut self, icon: impl Into<String>) -> Self {
        self.icon = Some(icon.into());
        self
    }

    /// Sets the column behavior
    pub fn with_behavior(mut self, behavior: ColumnBehavior) -> Self {
        self.behavior = Some(behavior);
        self
    }

    /// Sets the column groups for multi-state columns
    pub fn with_groups(mut self, groups: Vec<StateGroup>) -> Self {
        self.groups = Some(groups);
        self
    }
}

/// Behavior overrides for a workflow column
#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
pub struct ColumnBehavior {
    /// Skip review for tasks in this column
    #[serde(skip_serializing_if = "Option::is_none")]
    pub skip_review: Option<bool>,
    /// Auto-advance to next column when complete
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auto_advance: Option<bool>,
    /// Override agent profile for this column
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_profile: Option<String>,
}

impl ColumnBehavior {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_skip_review(mut self, skip: bool) -> Self {
        self.skip_review = Some(skip);
        self
    }

    pub fn with_auto_advance(mut self, advance: bool) -> Self {
        self.auto_advance = Some(advance);
        self
    }

    pub fn with_agent_profile(mut self, profile: impl Into<String>) -> Self {
        self.agent_profile = Some(profile.into());
        self
    }
}

/// Default agent profile configuration for a workflow
#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
pub struct WorkflowDefaults {
    /// Default worker agent profile
    #[serde(skip_serializing_if = "Option::is_none")]
    pub worker_profile: Option<String>,
    /// Default reviewer agent profile
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reviewer_profile: Option<String>,
}

/// External sync configuration (placeholder for future implementation)
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExternalSyncConfig {
    /// The external provider
    pub provider: SyncProvider,
    /// Status mapping from external to internal
    #[serde(default)]
    pub mapping: std::collections::HashMap<String, ExternalStatusMapping>,
    /// Sync settings
    pub sync: SyncSettings,
    /// How to resolve conflicts
    pub conflict_resolution: ConflictResolution,
}

/// Supported external sync providers
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SyncProvider {
    Jira,
    Github,
    Linear,
    Notion,
}

impl fmt::Display for SyncProvider {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SyncProvider::Jira => write!(f, "jira"),
            SyncProvider::Github => write!(f, "github"),
            SyncProvider::Linear => write!(f, "linear"),
            SyncProvider::Notion => write!(f, "notion"),
        }
    }
}

/// Mapping from an external status to internal status
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExternalStatusMapping {
    /// The external status name
    pub external_status: String,
    /// The internal status to map to
    pub internal_status: InternalStatus,
    /// The workflow column to display in
    pub column_id: String,
}

/// Sync direction settings
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SyncSettings {
    /// Sync direction
    pub direction: SyncDirection,
    /// Enable webhook for real-time sync
    #[serde(skip_serializing_if = "Option::is_none")]
    pub webhook: Option<bool>,
}

/// Sync direction options
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SyncDirection {
    Pull,
    Push,
    Bidirectional,
}

impl fmt::Display for SyncDirection {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SyncDirection::Pull => write!(f, "pull"),
            SyncDirection::Push => write!(f, "push"),
            SyncDirection::Bidirectional => write!(f, "bidirectional"),
        }
    }
}

/// Error for parsing SyncDirection from string
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParseSyncDirectionError {
    pub value: String,
}

impl fmt::Display for ParseSyncDirectionError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "unknown sync direction: '{}'", self.value)
    }
}

impl std::error::Error for ParseSyncDirectionError {}

impl FromStr for SyncDirection {
    type Err = ParseSyncDirectionError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "pull" => Ok(SyncDirection::Pull),
            "push" => Ok(SyncDirection::Push),
            "bidirectional" => Ok(SyncDirection::Bidirectional),
            _ => Err(ParseSyncDirectionError {
                value: s.to_string(),
            }),
        }
    }
}

/// Conflict resolution strategy
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ConflictResolution {
    ExternalWins,
    InternalWins,
    Manual,
}

impl fmt::Display for ConflictResolution {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ConflictResolution::ExternalWins => write!(f, "external_wins"),
            ConflictResolution::InternalWins => write!(f, "internal_wins"),
            ConflictResolution::Manual => write!(f, "manual"),
        }
    }
}

#[cfg(test)]
#[path = "workflow_tests.rs"]
mod tests;
